interfaces/connection.go

```diff
@@ -0,0 +1,250 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package interfaces
+
+import (
+	"fmt"
+
+	"github.com/snapcore/snapd/snap"
+)
+
+// Connection represents a connection between ConnectedPlug and ConnectedSlot
+type Connection struct {
+	plug *ConnectedPlug
+	slot *ConnectedSlot
+}
+
+// ConnectedPlug represents a plug that is connected to a slot.
+type ConnectedPlug struct {
+	plugInfo     *snap.PlugInfo
+	staticAttrs  map[string]interface{}
+	dynamicAttrs map[string]interface{}
+}
+
+// ConnectedSlot represents a slot that is connected to a plug.
+type ConnectedSlot struct {
+	slotInfo     *snap.SlotInfo
+	staticAttrs  map[string]interface{}
+	dynamicAttrs map[string]interface{}
+}
+
+// NewConnectedSlot creates an object representing a connected slot.
+func NewConnectedSlot(slot *snap.SlotInfo, dynamicAttrs map[string]interface{}) *ConnectedSlot {
+	return &ConnectedSlot{
+		slotInfo:     slot,
+		staticAttrs:  copyAttributes(slot.Attrs),
+		dynamicAttrs: dynamicAttrs,
+	}
+}
+
+// NewConnectedPlug creates an object representing a connected plug.
+func NewConnectedPlug(plug *snap.PlugInfo, dynamicAttrs map[string]interface{}) *ConnectedPlug {
+	return &ConnectedPlug{
+		plugInfo:     plug,
+		staticAttrs:  copyAttributes(plug.Attrs),
+		dynamicAttrs: dynamicAttrs,
+	}
+}
+
+// Interface returns the name of the interface for this plug.
+func (plug *ConnectedPlug) Interface() string {
+	return plug.plugInfo.Interface
+}
+
+// Name returns the name of this plug.
+func (plug *ConnectedPlug) Name() string {
+	return plug.plugInfo.Name
+}
+
+// Snap returns the snap Info of this plug.
+func (plug *ConnectedPlug) Snap() *snap.Info {
+	return plug.plugInfo.Snap
+}
+
+// Apps returns all the apps associated with this plug.
+func (plug *ConnectedPlug) Apps() map[string]*snap.AppInfo {
+	return plug.plugInfo.Apps
+}
+
+// Hooks returns all the hooks associated with this plug.
+func (plug *ConnectedPlug) Hooks() map[string]*snap.HookInfo {
+	return plug.plugInfo.Hooks
+}
+
+// SecurityTags returns the security tags for this plug.
+func (plug *ConnectedPlug) SecurityTags() []string {
+	return plug.plugInfo.SecurityTags()
+}
+
+// StaticAttr returns a static attribute with the given key, or error if attribute doesn't exist.
+func (plug *ConnectedPlug) StaticAttr(key string) (interface{}, error) {
+	if val, ok := plug.staticAttrs[key]; ok {
+		return val, nil
+	}
+	return nil, fmt.Errorf("attribute %q not found", key)
+}
+
+// StaticAttrs returns all static attributes.
+func (plug *ConnectedPlug) StaticAttrs() map[string]interface{} {
+	return plug.staticAttrs
```

Comment:
Ditto

Reply:
The constructor of ConnectedPlug/Slot creates a deep copy of static attributes, so modifications done directly to the map returned by this method will only affect this instance. Which would be nice to avoid, yes. But I'm not sure if we want a deep copy every time we return the map here?